Library/Homebrew/formula_versions.rb

```diff
@@ -15,56 +15,64 @@ class FormulaVersions
     ErrorDuringExecution, LoadError, MethodDeprecatedError
   ].freeze
 
-  MAX_VERSIONS_DEPTH = 2
-
-  attr_reader :name, :path, :repository, :entry_name
-
+  sig { params(formula: Formula).void }
   def initialize(formula)
     @name = formula.name
     @path = formula.path
-    @repository = formula.tap.path
-    @entry_name = @path.relative_path_from(repository).to_s
-    @current_formula = formula
+    @repository = T.must(formula.tap).path
+    @relative_path = @path.relative_path_from(repository).to_s
+    # Also look at e.g. older homebrew-core paths before sharding.
+    if (match = @relative_path.match(%r{^(HomebrewFormula|Formula)/([a-z]|lib)/(.+)}))
+      @old_relative_path = "#{match[1]}/#{match[3]}"
+    end
     @formula_at_revision = {}
   end
 
   def rev_list(branch)
     repository.cd do
       rev_list_cmd = ["git", "rev-list", "--abbrev-commit", "--remove-empty"]
-      rev_list_cmd << "--first-parent" if repository != CoreTap.instance.path
-      Utils.popen_read(*rev_list_cmd, branch, "--", entry_name) do |io|
-        yield io.readline.chomp until io.eof?
+      [relative_path, old_relative_path].compact.each do |entry|
+        Utils.popen_read(*rev_list_cmd, branch, "--", entry) do |io|
+          yield [io.readline.chomp, entry] until io.eof?
+        end
       end
     end
   end
 
-  sig { params(rev: String).returns(String) }
-  def file_contents_at_revision(rev)
-    repository.cd { Utils.popen_read("git", "cat-file", "blob", "#{rev}:#{entry_name}") }
-  end
-
   sig {
     type_parameters(:U)
-      .params(rev: String, _block: T.proc.params(arg0: Formula).returns(T.type_parameter(:U)))
-      .returns(T.nilable(T.type_parameter(:U)))
+      .params(
+        revision:              String,
+        formula_relative_path: String,
+        _block:                T.proc.params(arg0: Formula).returns(T.type_parameter(:U)),
+      ).returns(T.nilable(T.type_parameter(:U)))
   }
-  def formula_at_revision(rev, &_block)
+  def formula_at_revision(revision, formula_relative_path = relative_path, &_block)
```

Comment:
I have a vague memory of using this in `test-bot`. I'll double check later, but it would be good to fix any callers there too.

Reply:
I don't see any uses outside of `brew`.